src/app/CASESessionManager.h

```diff
@@ -87,6 +87,60 @@ class CASESessionManager : public OperationalSessionReleaseDelegate, public Sess
 #endif // CHIP_DEVICE_CONFIG_ENABLE_AUTOMATIC_CASE_RETRIES
     );
 
+    /**
+     * Find an existing session for the given node ID or trigger a new session request.
+     *
+     * The caller can optionally provide `onConnection` and `onSetupFailure`
+     * callback objects. If provided, these will be used to inform the caller about successful or
+     * failed connection establishment.
+     *
+     * If the connection is already established, the `onConnection` callback will be immediately called,
+     * before `FindOrEstablishSession` returns.
+     *
+     * The `onSetupFailure` callback may be called before the `FindOrEstablishSession`
+     * call returns, for error cases that are detected synchronously.
+     *
+     * The `attemptCount` parameter can be used to automatically retry multiple times if session setup is
+     * not successful.
+     *
+     * @param peerId The node ID to find or establish a session with.
+     * @param onConnection A callback to be called upon successful connection establishment.
+     * @param onSetupFailure A callback to be called upon an extended device connection failure.
+     * @param attemptCount The number of retry attempts if session setup fails (default is 1).
+     * @param onRetry A callback to be called on a retry attempt (enabled by a config flag).
+     */
+    void FindOrEstablishSession(const ScopedNodeId & peerId, Callback::Callback<OnDeviceConnected> * onConnection,
+                                Callback::Callback<OperationalSessionSetup::OnSetupFailure> * onSetupFailure
+#if CHIP_DEVICE_CONFIG_ENABLE_AUTOMATIC_CASE_RETRIES
+                                ,
+                                uint8_t attemptCount = 1, Callback::Callback<OnDeviceConnectionRetry> * onRetry = nullptr
+#endif // CHIP_DEVICE_CONFIG_ENABLE_AUTOMATIC_CASE_RETRIES
+    );
+
+    /**
+     * Find an existing session for the given node ID or trigger a new session request.
+     *
+     * The caller can optionally provide `onConnection`
+     * callback objects. If provided, these will be used to inform the caller about successful connection establishment.
+     *
+     * If the connection is already established, the `onConnection` callback will be immediately called,
+     * before `FindOrEstablishSession` returns.
+     *
+     * The `attemptCount` parameter can be used to automatically retry multiple times if session setup is
+     * not successful.
+     *
+     * @param peerId The node ID to find or establish a session with.
+     * @param onConnection A callback to be called upon successful connection establishment.
+     * @param attemptCount The number of retry attempts if session setup fails (default is 1).
+     * @param onRetry A callback to be called on a retry attempt (enabled by a config flag).
+     */
+    void FindOrEstablishSession(const ScopedNodeId & peerId, Callback::Callback<OnDeviceConnected> * onConnection, nullptr_t
```

Comment:
The reasonable person will ask why this thing is here at all.  The answer is: "to make passing nullptr for the error handler compile", and should probably be documented.